### DISCUSSION TITLE
vacuum_restore_map_from_file - how do I replace the saved map in recovery_map_list ?

### DISCUSSION THREAD

**Discussion by gibman** [Tasshack/dreame-vacuum#515] [Q&A] [answered]
been playing around with the service call "vacuum_restore_map_from_file"  to see how it affects the map system on the vacuum robot.

I started by deleting all maps from within the dreame app then created a new map scan job.
I had to restart the dreame app integration manually in order to have the changes reflected in the camera.xyz entities. the map_id did not bump up. Once I restarted the integration the map_id went from eg. 14 to 15.
Maybe if I wait long enough it will update ?

It looks like when succesfully calling the service "vacuum_restore_map_from_file" the map change is reflected on the "main" map page on the dreame app itself.

What I mean is if I goto the map edit page on the dreame app, then another older map is presented and not the newly uloaded one I just did.
So once I go down the route with calling vacuum_restore_map_from_file, I have no way of editing this map any further as it will not appear in the list of maps in "restore maps" from the edit map section on the dreame app.

When I look at the camera.vacuum entity I can see the saved maps are listed as under the dictionary: recovery_map_list
So I guess there is no way to replace those currently ?

Am I missing something here ?

Although I was able to recover an older perfect golden map on my newly factory resat vacuum - how ??
I can do this but I will not be able to edit it any further.

If you want the steps for this they are:
Reset vacuum (then map_id will be 0) or delete all maps (if you delete the maps, it will bump the map index +1).
initiate a new vacuum map from the dreame. I doesnt have to be very good. Meaning it just needs to create an initial map.
No need to clean up the house like you did the very first time..
(God knows I would die for an offline map editor that could act as an eraser tool for the stuff in the room that I dont want to be part of the map. Currently I have to fool the system by putting away chairs etc. in order to get clean rooms without any obstacles.)

Then go into map edit on the dreame app, then create an extra backup.
Once it has completed this then do a map download from your browser.
eg. /api/camera_recovery_map_proxy/{entity_id}?token={access_token}&index={recovery_map_index}&file={download_file}&info={info_text}
as described here: https://github.com/Tasshack/dreame-vacuum/blob/dev/docs/map.md

you will use this map as a base for the modification.

now open the initial tbz2 in 7zip (windows).
now also open the original golden map (the map you want to apply on your vacuum)

the initial tbz2 contents is:  (15 is map_id, it will vary in your case)
15 (folder)
divide/15 (folder)
15.dat (file)

 original golden map content is:
1 (folder)
divide/1 (folder)
1.dat (file)

you will need to modify the initial tbz2 file (do it on a new file copy eg.)
1) copy the content of folder "1" of the golden map tbz2 file to the folder named "15" in the initial tbz2 archive.
2) copy the content of folder "divide/1" of the golden map tbz2 file to the folder named "divide/15" in the initial tbz2 archive.
3) delete the 15.dat file in the initial tbz2 archive.
4) copy 1.dat file from the golden tbz2 archive to the root of the initial tbz2 archive.
5) rename 1.dat to 15.dat on the initial tbz2 archive.

in other words:
copy golden/1/*.* => init/15/
copy golden/divide/1/*.* => init/devide/15/
del init/15.dat
copy golden/divide/1.dat => init/devide/15.dat

then upload the modified file to the vacuum  using ..
dreame_vacuum.vacuum_restore_map_from_file
see docs.

doing these steps and I am able to have my old map even through a fact. reset procedure.
although I am unable to have the map appear in the map editing section of the dreame app thus I cant improve upon it. 
eg by adding new no go zones etc.

EDIT: _I just found a useful workaround for having the newly uploaded locally stored map appear in the map editing section of the dreame app.
When you open the edit map section on your dreame app, you will probably see the basic incomplete initial map.
This is fine as you just have to rename one of the rooms on the map.
Once you rename a room to anything you want to then you will see the newly upload file being applied. This is the workaround.. It's that simple.
WOW at last.
Now complete the rename or an other edits you may wish to do. Finally click save backup. The map should now appear in the dreame app as a backed up map. Which again means that you can edit it again and again in the future should you wish to.
Just remember to DOWNLOAD it again to your home assistant folder :)_
Also I just checked, the map_id. At this point it is still 15.

vacuum bot: l10s ultra
dreame integration: beta 13

**Comment by Tasshack** [comment]
Is there anything I can do about this?

**Comment by gibman** [comment]
sorry for the late reply.

Well I did seem to conclude that the "map_id" + the the other attributes related to downloading the map on the camera.xyz entity did not reflect the actual state. This happened when I did changes to the map in the official Dreame android app.
The old map_id etc. was still being shown.

Easy to repro. just create a new backup map in the dreame app.

They way I circumvented that was merely to reload the Dreame hass integration in hass. Then all the data attributes of the camera entity would be spot on and actual.

Onto the actual case itself...
Have you been able to take an older tbz2 archive and re-apply this succesfully on a factory resat vacuum or maybe just a situation where you have deleted all maps thus starting over from scratch ?

If we could somehow remove the mundane manual steps involved with preparing an old map archive that the enduser would want to reapply on a vacuum setup that now has a non matching map_id ?

I guess your integration could take care of this by applying hese changes on the fly on the archive itself in order to artificially keep the map_id of the archive 100% in sync with the actual current map_id of the vacuum.

It's just a matter of renaming the previously named folders etc. within the archive
The only file I did not manage to figure out was the "<map_id>.dat" file.
I would need to re-download the recovery map once I have a new map_id on the vacuum.
Then grab the .dat file, rename it to the current map_id, then add it to the archive.

I tried comparing the older dat file to the new one in the most recent archive.
There were changes but I couldnt see a pattern where I could programmatically create my own copy through code :)

I guess this could be automated by your integration as well.

Atleast I have the manual steps on how to go about.
If you need me to clarify anything, then just let me know :)
I know it's rather confusing.

**Reply by Tasshack** [reply, marked as answer]
I won't add anything to the integration can potentialy break your device. Modifying a packaged achive and uploading to device a modified store file can result a boot loop and potentially soft brick your device. These devices are really built to be made to be at low cost that means software is very badly implemented and there is no checks for unexpected values in anywhere even api requests. Also you cannot expect this can be a thing tested by any engineer who built the device so if anything goes wrong you will probably have a very expensive brick. Not to mentioning there are more than 5 generation of devices which has different capabilities related to the map data so your solution might not work on my older or newer device.

I can only add an action needs to be done to the integration after the modified recovery map has been successfully uploaded and loaded as current map to be saved to the saved map list so I need the information about what have you actually done to map to be listed on the saved map list. Because anything that you did from the app to save the modified recovery map can also be safetly done from the integration and won't break anything (probably) so I can add that part of the automation to the project.
